File: packages/lib-angular/projects/avatars/src/lib/avatars/avatar-sunset.ts
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";

import { Component, Input } from "@angular/core";

import type { AvatarProps } from "../avatar.utils";

import { defaultAvatarProps } from "../avatar.utils";
import { generateColors, SIZE } from "./avatar-sunset.utils";

@Component({
  selector: "avatar-sunset, AvatarSunset",
  template: `
    <svg
      fill="none"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      [attr.viewBox]='"0 0 " + SIZE + " " + SIZE'
      [attr.width]="size"
      [attr.height]="size"
    >
      <ng-container *ngIf="hasTitle">
        <title>{{name}}</title>
      </ng-container>

      <mask
        id="mask__sunset"
        maskUnits="userSpaceOnUse"
        [attr.x]="0"
        [attr.y]="0"
        [attr.width]="SIZE"
        [attr.height]="SIZE"
      >
        <rect
          fill="#FFFFFF"
          [attr.width]="SIZE"
          [attr.height]="SIZE"
          [attr.rx]="square ? undefined : SIZE * 2"
        ></rect>
      </mask>

      <g mask="url(#mask__sunset)">
        <path d="M0 0h80v40H0z" [attr.fill]="getPaintFill0()"></path>

        <path d="M0 40h80v40H0z" [attr.fill]="getPaintFill1()"></path>
      </g>

      <defs>
        <linearGradient
          gradientUnits="userSpaceOnUse"
          [attr.id]="getPaintId0()"
          [attr.x1]="SIZE / 2"
          [attr.y1]="0"
          [attr.x2]="SIZE / 2"
          [attr.y2]="SIZE / 2"
        >
          <stop [attr.stop-color]="sunsetColors(name, colors)[0]"></stop>

          <stop
            [attr.offset]="1"
            [attr.stop-color]="sunsetColors(name, colors)[1]"
          ></stop>
        </linearGradient>

        <linearGradient
          gradientUnits="userSpaceOnUse"
          [attr.id]="getPaintId1()"
          [attr.x1]="SIZE / 2"
          [attr.y1]="SIZE / 2"
          [attr.x2]="SIZE / 2"
          [attr.y2]="SIZE"
        >
          <stop [attr.stop-color]="sunsetColors(name, colors)[2]"></stop>

          <stop
            [attr.offset]="1"
            [attr.stop-color]="sunsetColors(name, colors)[3]"
          ></stop>
        </linearGradient>
      </defs>
    </svg>
  `,
})
export class AvatarSunset {
  SIZE = SIZE;

  @Input() name: Omit<AvatarProps, 'variant'>['name'] = defaultAvatarProps['name'];
  @Input() size: Omit<AvatarProps, 'variant'>['size'] = defaultAvatarProps['size'];
  @Input() hasTitle: Omit<AvatarProps, 'variant'>['hasTitle'] = defaultAvatarProps['hasTitle'];
  @Input() square: Omit<AvatarProps, 'variant'>['square'] = defaultAvatarProps['square'];
  @Input() colors: Omit<AvatarProps, 'variant'>['colors'] = defaultAvatarProps['colors'];

  get formattedName() {
    return this.name.replace(/\s/g, "");
  }
  sunsetColors(name, colors) {
    return generateColors(name, colors);
  }
  getPaintFill0() {
    return "url(#gradient_paint0_linear_" + this.formattedName + ")";
  }
  getPaintFill1() {
    return "url(#gradient_paint1_linear_" + this.formattedName + ")";
  }
  getPaintId0() {
    return "gradient_paint0_linear_" + this.formattedName;
  }
  getPaintId1() {
    return "gradient_paint1_linear_" + this.formattedName;
  }
}

@NgModule({
  declarations: [AvatarSunset],
  imports: [CommonModule],
  exports: [AvatarSunset],
})
export class AvatarSunsetModule {}

File: packages/lib-leptos/src/avatar_sunset.rs
use crate::avatar_sunset_utils::*;
use crate::avatar_utils::*;
use leptos::*;

#[component]
pub fn AvatarSunset(
    cx: Scope,
    name: String,
    colors: Vec<&'static str>,
    square: bool,
) -> impl IntoView {
    let num_from_name = hash_code(&name);
    let generated_colors = generate_colors(&name, colors);

    let HALF_SIZE = SIZE / 2;
    let outer_radius = if square == true { 0.0 } else { (SIZE as f64) * 2.0 };

    view! { cx,
      <svg
        fill="none"
        role="img"
        xmlns="http://www.w3.org/2000/svg"
        viewBox=format!("0 0 {} {}",  SIZE, SIZE)
        width=200
        height=200
      >
        <mask
        id=format!("mask__sunset_{}", num_from_name)
          maskUnits="userSpaceOnUse"
          x=0
          y=0
          width=SIZE
          height=SIZE
        >
          <rect
            fill="#FFFFFF"
            width=SIZE
            height=SIZE
            rx=outer_radius
          />
        </mask>

        <g mask=format!("url(#mask__sunset_{})", num_from_name)>
          <path d="M0 0h80v40H0z" fill=format!("url(#gradient_paint0_linear_{})", num_from_name) />
          <path d="M0 40h80v40H0z" fill=format!("url(#gradient_paint1_linear_{})", num_from_name) />
        </g>

        <defs>
          <linearGradient
            gradientUnits="userSpaceOnUse"
            id=format!("gradient_paint0_linear_{}", num_from_name)
            x1=HALF_SIZE
            y1=0
            x2=HALF_SIZE
            y2=HALF_SIZE
          >
            <stop stop-color=generated_colors[0].clone() />

            <stop
              offset=1
              stop-color=generated_colors[1].clone()
            />
          </linearGradient>

          <linearGradient
            gradientUnits="userSpaceOnUse"
            id=format!("gradient_paint1_linear_{}", num_from_name)
            x1=HALF_SIZE
            y1=HALF_SIZE
            x2=HALF_SIZE
            y2=SIZE
          >
            <stop stop-color=generated_colors[2].clone() />

            <stop
              offset=1
              stop-color=generated_colors[3].clone()
            />
          </linearGradient>
        </defs>


      </svg>
    }
}
